In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('Real estate.csv')
df.head()

No  X1 transaction date  X2 house age  \
0   1             2012.917          32.0   
1   2             2012.917          19.5   
2   3             2013.583          13.3   
3   4             2013.500          13.3   
4   5             2012.833           5.0   

   X3 distance to the nearest MRT station  X4 number of convenience stores  \
0                                84.87882                               10   
1                               306.59470                                9   
2                               561.98450                                5   
3                               561.98450                                5   
4                               390.56840                                5   

   X5 latitude  X6 longitude  Y house price of unit area  
0     24.98298     121.54024                        37.9  
1     24.98034     121.53951                        42.2  
2     24.98746     121.54391                        47.3  
3     24.98746     121.54391                        54.8  
4     24.97937     121.54245                        43.1

In [3]:
df.dtypes

No                                          int64
X1 transaction date                       float64
X2 house age                              float64
X3 distance to the nearest MRT station    float64
X4 number of convenience stores             int64
X5 latitude                               float64
X6 longitude                              float64
Y house price of unit area                float64
dtype: object

In [4]:
# Checking wether is missing values are not
df.isna().sum()

No                                        0
X1 transaction date                       0
X2 house age                              0
X3 distance to the nearest MRT station    0
X4 number of convenience stores           0
X5 latitude                               0
X6 longitude                              0
Y house price of unit area                0
dtype: int64

In [5]:
df.shape

(414, 8)

## Droping first column
### Becouse its just a serial no we already have defult serial no so we dont need extra serial no.

In [6]:
df.columns

Index(['No', 'X1 transaction date', 'X2 house age',
       'X3 distance to the nearest MRT station',
       'X4 number of convenience stores', 'X5 latitude', 'X6 longitude',
       'Y house price of unit area'],
      dtype='object')

In [7]:
df_new = df.drop('No', axis = 1)
df_new

X1 transaction date  X2 house age  \
0               2012.917          32.0   
1               2012.917          19.5   
2               2013.583          13.3   
3               2013.500          13.3   
4               2012.833           5.0   
..                   ...           ...   
409             2013.000          13.7   
410             2012.667           5.6   
411             2013.250          18.8   
412             2013.000           8.1   
413             2013.500           6.5   

     X3 distance to the nearest MRT station  X4 number of convenience stores  \
0                                  84.87882                               10   
1                                 306.59470                                9   
2                                 561.98450                                5   
3                                 561.98450                                5   
4                                 390.56840                                5   
..                                      ...                              ...   
409                              4082.01500                                0   
410                                90.45606                                9   
411                               390.96960                                7   
412                               104.81010                                5   
413                                90.45606                                9   

     X5 latitude  X6 longitude  Y house price of unit area  
0       24.98298     121.54024                        37.9  
1       24.98034     121.53951                        42.2  
2       24.98746     121.54391                        47.3  
3       24.98746     121.54391                        54.8  
4       24.97937     121.54245                        43.1  
..           ...           ...                         ...  
409     24.94155     121.50381                        15.4  
410     24.97433     121.54310                        50.0  
411     24.97923     121.53986                        40.6  
412     24.96674     121.54067                        52.5  
413     24.97433     121.54310                        63.9  

[414 rows x 7 columns]

In [8]:
df_new.shape

(414, 7)

## Extract Train dataset and Test dataset
### 2. Separate out independent and target variables

In [9]:
X = df_new.drop('Y house price of unit area', axis=1)
y = df_new.loc[:, 'Y house price of unit area']

X.shape, y.shape

((414, 6), (414,))

### 3. Split the dataset into train and test with 25% data in test
### Can be easily established using sklearn
#### Usually go for 70-30 split

#### or 80-20

#### or 75-25

In [10]:
from sklearn.model_selection import train_test_split

In [11]:
X_train, X_test, y_train, y_test = train_test_split(
                                                      X, 
                                                      y, 
                                                      test_size=0.30, 
                                                      random_state=1234
                                                   )

X_train.shape, y_train.shape, X_test.shape, y_test.shape

((289, 6), (289,), (125, 6), (125,))

## Linear Regression Model

#### => Data should always be 2 dimensional
#### => Create a Linear Regression Model object
#### => Fit the train data to create co-effecients from the training data

In [12]:
from sklearn.linear_model import LinearRegression

In [13]:
lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

LinearRegression()

In [14]:
y_learnt = lr_model.predict(X_train)

In [15]:
y_pred = lr_model.predict(X_test)

In [16]:
y_pred[:5]

array([39.37931134, 50.85803606, 33.98509832, 40.99405457, 41.72026962])

In [17]:
from sklearn.metrics import mean_squared_error as mse

In [18]:
rmse = lambda actual,pred : np.sqrt(mse(actual, pred))

In [19]:
rmse_train = rmse(y_train, y_learnt)
rmse_test  = rmse(y_test, y_pred)

rmse_train, rmse_test

(8.09790885285676, 10.268560291657867)

# One importent note is 
### Our test rmse is always less then train rmse
### Then only r2_score will increase other wise r2_score wil decrease
### if r2 increases our model will be good
### if r2 decreases our model pridictions will be not good

## some examples are

### r2_1 = 1 - (8/10)    : 0.1999
### r2_2 = 1 - (6/10)    : 0.40
### r2_3 = 1 - (4/10)    : 0.60

In [20]:
from sklearn.metrics import r2_score

In [21]:
r2_score = r2_score(y_test, y_pred)
r2_score.round(4)

0.5039

In [22]:
print(r2_score.round(3) * 100,'%','of accuracy of our model')

50.4 % of accuracy of our model


### It means 50.4 % of accuracy of our model
### Still we need to improve our model now its only 50% better we need increase it to 90% then only our model will predict well
#### Train_rmse = 8.09
#### Test_rmse  = 10.26
#### Diffence is =  2.17 

### Adjusted r2 formula
### r2_adjusted = 1 - ( ( 1 - r2 ) * ( n - 1 ) / n - p - 1 )

### Here we all know that is when r2_score will decrease the adjusted r2 will also decrese
### r2 adjusted is mesure in between 0 to 1
### so r2_adjusted is near to 1 the model will become good
### so r2_adjusted is near to 0 the model will become bad
### Example : 

In [23]:
r2_score = r2_score.round(4)
r2_score

0.5039

In [24]:
num_of_predicters = len(y_pred)
num_of_predicters

125

In [25]:
Total_sample_size = len(y_learnt)
Total_sample_size

289

In [26]:
r2_adjusted = 1 - ((1 - r2_score)*(Total_sample_size - 1)/(Total_sample_size - num_of_predicters - 1))
r2_adjusted

0.1234552147239264

In [33]:
r2_1 = 1 - (8/10)
r2_1

0.19999999999999996

In [28]:
r2_2 = 1 - (6/10)
r2_2

0.4

In [29]:
r2_3 = 1 - (4/10)
r2_3

0.6